graphql/src/Mu/GraphQL/Quasi.hs

```diff
@@ -0,0 +1,148 @@
+{-# language DataKinds         #-}
+{-# language OverloadedStrings #-}
+{-# language TemplateHaskell   #-}
+{-# language ViewPatterns      #-}
+
+module Mu.GraphQL.Quasi where
+
+import qualified Data.Text                     as T
+import           Language.GraphQL.Draft.Parser (parseSchemaDoc)
+import qualified Language.GraphQL.Draft.Syntax as GQL
+import           Language.Haskell.TH
+
+import           Mu.Rpc
+import           Mu.Schema.Definition          ()
+
+-- | Imports an graphql definition written in-line as a 'Schema'.
+graphql :: String -> String -> T.Text -> Q [Dec]
+graphql scName svName schema =
+  case parseSchemaDoc schema of
+    Left e  -> fail ("could not parse graphql spec: " ++ show e)
+    Right p -> graphqlToDecls scName svName p
+
+data Result =
+    GQLSchema Type
+  | GQLService Type
+  deriving (Show, Eq)
+
+graphqlToDecls :: String -> String -> GQL.SchemaDocument -> Q [Dec]
+graphqlToDecls schemaName serviceName (GQL.SchemaDocument types) = do
+  let schemaName'  = mkName schemaName
+      serviceName' = mkName serviceName
+  rs <- traverse (typeToDec schemaName') types
+  let schemaTypes  = [x | GQLSchema  x <- rs]
+      serviceTypes = [x | GQLService x <- rs]
+  schemaDec <- tySynD schemaName' [] (pure $ typesToList schemaTypes)
+  serviceDec <- tySynD serviceName' [] (pure $ typesToList serviceTypes)
+  pure [schemaDec, serviceDec]
+
+typeToDec :: Name -> GQL.TypeDefinition -> Q Result
+typeToDec _ (GQL.TypeDefinitionScalar _)             = error "not implemented" -- TODO: handle "well-known scalars"
+typeToDec schemaName (GQL.TypeDefinitionObject objs) = objToDec objs
+  where
+    objToDec :: GQL.ObjectTypeDefinition -> Q Result
+    objToDec (GQL.ObjectTypeDefinition _ nm _ _ flds) =
+      GQLService <$> [t| 'Service $(textToStrLit $ GQL.unName nm) '[]
+          $(typesToList <$> traverse gqlFieldToType flds) |]
+    gqlFieldToType :: GQL.FieldDefinition -> Q Type
+    gqlFieldToType (GQL.FieldDefinition _ (GQL.unName -> fnm) args ftyp _) =
+      [t| 'Method $(textToStrLit fnm) '[]
+            $(typesToList <$> traverse argToType args)
+            'RetSingle $(retToType ftyp) |]
+    argToType :: GQL.InputValueDefinition -> Q Type
+    argToType (GQL.InputValueDefinition _ aname _ _) =
+      [t| 'ArgSingle 'Nothing '[] ('SchemaRef $(conT schemaName) $(textToStrLit (GQL.unName aname))) |]
```

Comment:
This is not correct, in several respects:

- if `aname` is the name of the argument, it should go as `Just $(textToStrLit (GQL.unName aname))` instad of `Nothing`,
- the last part should be the **type** of the argument. But it cannot be always `SchemaRef`, you have to take into account that it could be a list (to wrap in a `ListRef`), whether it's optional (to use `OptionalRef`), or whether it's a scalar (to use `TPrimitive` or give an error).

Reply:
I think I've fixed that, but now I don't have any reference to `'SchemaRef schemaName` 🤔 where should that happen?